### ISSUE TITLE
ImageCapture API

### ISSUE THREAD

**Issue by charlielee** [charlielee/boats-animator#205]
Recently I came across the `ImageCapture` API, which is still yet to be fully implemented in Chrome but looks pretty interesting.

It allows you to "properly" take a photo from a connected camera rather than just grab a frame from a video feed (as we currently do). In addition you'll eventually be able to set loads of different settings for the camera such as its focus, exposure, brightness, contrast etc. There are also `imageHeight` and `imageWidth` settings so I assume this would allow us to capture images at higher resolutions than some cameras support over live streaming (eg my Logitech c920 webcam supports 1080p but only streams up to 720p when called from `getUserMedia`).

I think it will probably be a while until we can consider implementing this into Boats Animator but I am really excited to have discovered this! 😄

#### Further information:
https://developers.google.com/web/updates/2016/12/imagecapture
https://w3c.github.io/mediacapture-image/
https://github.com/w3c/mediacapture-image/blob/gh-pages/implementation-status.md

**Comment by le717** [comment]
Ooooo, interesting! We'll have to keep this in mind for the future! 😃

**Comment by BoatsAreRockable** [comment]
@le717 things seem to be moving along pretty quickly with this! 
https://github.com/w3c/mediacapture-image/blob/gh-pages/implementation-status.md

I have just come back from holiday and was going to give #207 another review before potentially merging it. But now I don't know whether it needs completely rethinking again.

**Comment by le717** [comment]
I haven't forgot about this! When I saw the `ImageCapture` API was out, my first thought was Boats. I'm not sure right now what all will be impacted by it. ~~I know we'll still need `getUserMedia` for the video stream and camera selection, but as for the rest, we'll need to explore. `ImageCapture` permits pictures in the highest resolution, but we'll still need to provide an option for a specific resolution (defaulting to the highest). ~~

I just looked at the Google Developers blog post again. We'll still need a lot of the WebRTC code already written (which I'll eventually get to look at!) but implementing `ImageCapture` may be more straight-forward than I initially thought.

**Comment by le717** [comment]
This might be something to look back into with all the requests for built-in camera controls (#240) coming in. This API supports changing camera settings, like exposure and brightness (at least in Chrome and per spec, [1](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getSupportedConstraints), [2](https://w3c.github.io/mediacapture-image/)). It may also help with app performance (#238) through the [`grabFrame()`](https://developer.mozilla.org/en-US/docs/Web/API/ImageCapture/grabFrame) and [`takePhoto()`](https://developer.mozilla.org/en-US/docs/Web/API/ImageCapture/takePhoto) APIs.

**Comment by charlielee** [comment]
I definitely want to use the `takePhoto()` API so we actually capture camera exposures rather than frames from the stream.

Annoyingly, I don't think we'll be able to use MediaStreamTrack settings for camera controls, at least not fully. Linux and Windows seem to be missing a few controls, and **on macOS basically none have been implemented**. We need as complete manual controls as possible.

It may make sense to use a hybrid approach where the Web APIs are used where possible, but if a control is missing then a lower level library is used as a fallback.

In the past few weeks, I've been experimenting with the accessing Windows camera API itself and made some good progress. My current plan is to create a C# library / CLI to access this. This is will then be accessed by Boats Animator either with [electron-edge-js](https://github.com/agracio/electron-edge-js) or just using the CLI. I've made a decent amount of progress and will have more to share in the coming weeks.

For macOS/Linux, I'm hoping we can use [node-uvc-control](https://github.com/makenai/node-uvc-control). Otherwise, some possible alternatives are forking [uvc-util](https://github.com/jtfrey/uvc-util) for macOS and doing something with [v4l2-ctl](https://www.mankier.com/1/v4l2-ctl) for Linux.



I've tried to match the native and Web API controls that are listed for my Logitech c920 on Windows to see what's missing from the Web APIs.

| Windows Native Control     | Value | MediaStreamTrack Control | Value  |
| -------------------------- | --------- | ------------------------ | --------- |
| Brightness                 | 128       | brightness               | 128       |
| Contrast                   | 128       | contrast                 | 128       |
| Saturation                 | 128       | saturation               | 128       |
| Sharpness                  | 128       | sharpness                | 128       |
| White Balance              | 2808      | colorTemperature         | 2842      |
| Auto White Balance         | false     | whiteBalanceMode         | "manual"  |
| **Backlight Comp**         | **0**     | **!!!**                  | **!!!**   |
| Gain                       | 77        | exposureCompensation     | 76        |
| **PowerLine Frequency**    | **60 Hz** | **!!!**                  | **!!!**   |
| Zoom                       | 100       | zoom                     | 100       |
| Focus                      | 0         | focusDistance            | 0         |
| Auto Focus                 | false     | focusMode                | "manual"  |
| **Exposure**               | **-5**    | **!!!**                  | **!!!**   |
| Auto Exposure              | false     | exposureMode             | "manual"  |
| Pan                        | 0         | pan                      | 0         |
| Tilt                       | 0         | tilt                     | 0         |
| **Low Light Compensation** | **false** | **!!!**                  | **!!!**   |
| **!!!**                    | !!!       | **exposureTime** \*         | **312.5** |
| *Hue*                      | *N/A*     |                          |           |
| *Gamma*                    | *N/A*     |                          |           |
| *ColorEnable*              | *N/A*     |                          |           |
| *Aperture (Iris)*          | *N/A*     |                          |           |
| *Roll*                     | *N/A*     |                          |           |

\* I assume the Web API `exposureTime` value correlates to the native `Exposure` value somehow.

### Windows Built In Control Window

![c920 Windows Settings 1](https://user-images.githubusercontent.com/8095888/112900024-c9edb080-90da-11eb-89e5-18275987f283.png)
![c920 Windows Settings 2](https://user-images.githubusercontent.com/8095888/112900032-cbb77400-90da-11eb-8ee2-696efaca69c2.png)

### Accessing MediaStreamTrack / ImageCapture Settings

```javascript
navigator.mediaDevices.enumerateDevices().then((d) => console.log(d) );

navigator.mediaDevices.getUserMedia({
  video: true
})
.then(async mediaStream => {
  let track = mediaStream.getVideoTracks()[0];
  let imageCapture = new ImageCapture(track);
  
  console.log(track.label);
  console.log(track.getCapabilities());
  console.log(track.getSettings());

  let capabilities = await imageCapture.getPhotoCapabilities()
  console.log(capabilities);

  let settings = await imageCapture.getPhotoSettings();
  console.log(settings);
})
.catch(error => console.log(error));
```

![c920 Windows MediaStreamTrack](https://user-images.githubusercontent.com/8095888/112900037-cd813780-90da-11eb-90d2-cd07128e981a.png)

**Comment by charlielee** [comment]
### macOS MediaStreamTrack settings

Here's the same code for a Logitech c920 run on macOS Catalina:

![image](https://user-images.githubusercontent.com/8095888/112905839-0cb38680-90e3-11eb-93dd-ee6d14045f0e.png)

**Comment by charlielee** [comment]
Closed as v1.0 utilises this API
